### PR TITLE
PSBTv2 (BIP-370) support and PSBT hardening

### LINKED ISSUE #65: Invalid Taproot signatures in transactions with non-standard sighash
Apparently embit v0.8.0 produces incorrect signatures in Taproot transactions with a non-standard sighash.

The following script reproduces the issue. The unsigned PSBT has been prepared in Sparrow with SINGLE|ANYONECANPAY. We call sign_with() with `None` as the sighash so that the sighash defined in the psbt data is the one taken into account.

```python
from base64 import b64decode
from embit.bip32 import HDKey
from embit.psbt import PSBT

# tprv of "aim" x12 in testnet/signet
master_private_key = 'tprv8ZgxMBicQKsPf7HFb4WqgmtZCLyuWp3qdDDMngGZUcVLK27uovLZmq87VQRzmRLNGGfuv9MmbrfDfFqYWN9UC4jU9yFPFBKznKtjJXAp6zB'
unsigned_psbt_base64 = 'cHNidP8BAF4CAAAAARfCyWIG/NiHOFrFkb34/tlVFc/N0wVIGQzWQrx0j2pLAAAAAAD9////AbBjzR0AAAAAIlEgl3d2EP+VhSPaLqFAcO3gOmaApwdP0ulTfjbhZnQkGpZNPgAATwEENYfPA6fJqKyAAAAANFK5sPQqTwnjADO2lv7oWYdTo0CWEwxtTtGDWvPUiMsD1hM7Pe2Gzg01QefANL7alLQYwxsmeyTFfH/TlstresUQnRILGVYAAIABAACAAAAAgAABAF4CAAAAARIkR5wC2SeGdLpgweIuxsxD3wICJsUWtJ3T1ZIrxsyjAAAAAAACAAAAASBkzR0AAAAAIlEg0vcpFkr36+uyRermS3owVW2WZAoX2RxiCnLL4jFfaCUAAAAAAQErIGTNHQAAAAAiUSDS9ykWSvfr67JF6uZLejBVbZZkChfZHGIKcsviMV9oJQEDBIMAAAAhFmNl9RJbp6Iz+FwX7WJG3TcKWmrvBr1tvpdVJ9n6XGUTGQCdEgsZVgAAgAEAAIAAAACAAAAAAAIAAAABFyBjZfUSW6eiM/hcF+1iRt03Clpq7wa9bb6XVSfZ+lxlEwAhByx1NQ1R+Sx4GcdHzkgnsRJh7SapmaqPBXiewFXgL3RdGQCdEgsZVgAAgAEAAIAAAACAAAAAAAMAAAABBSAsdTUNUfkseBnHR85IJ7ESYe0mqZmqjwV4nsBV4C90XQA='

xkey = HDKey.from_base58(master_private_key)

unsigned_psbt_raw = b64decode(unsigned_psbt_base64)

psbt = PSBT.parse(unsigned_psbt_raw)

sigs = psbt.sign_with(xkey, None)

print(sigs)
print(psbt.to_base64())
```

Output:
```
1
cHNidP8BAF4CAAAAARfCyWIG/NiHOFrFkb34/tlVFc/N0wVIGQzWQrx0j2pLAAAAAAD9////AbBjzR0AAAAAIlEgl3d2EP+VhSPaLqFAcO3gOmaApwdP0ulTfjbhZnQkGpZNPgAATwEENYfPA6fJqKyAAAAANFK5sPQqTwnjADO2lv7oWYdTo0CWEwxtTtGDWvPUiMsD1hM7Pe2Gzg01QefANL7alLQYwxsmeyTFfH/TlstresUQnRILGVYAAIABAACAAAAAgAABAF4CAAAAARIkR5wC2SeGdLpgweIuxsxD3wICJsUWtJ3T1ZIrxsyjAAAAAAACAAAAASBkzR0AAAAAIlEg0vcpFkr36+uyRermS3owVW2WZAoX2RxiCnLL4jFfaCUAAAAAAQErIGTNHQAAAAAiUSDS9ykWSvfr67JF6uZLejBVbZZkChfZHGIKcsviMV9oJQEDBIMAAAABCEMBQYSwoL7UpZlJvIsjQLxnFlxNdvxCFsxWQbWaKyfEgPCLZzzejYIl4gIo/LVYx+z48A0U3GxSO/5nth8ExBIJbUWDIRZjZfUSW6eiM/hcF+1iRt03Clpq7wa9bb6XVSfZ+lxlExkAnRILGVYAAIABAACAAAAAgAAAAAACAAAAARcgY2X1ElunojP4XBftYkbdNwpaau8GvW2+l1Un2fpcZRMAAQUgLHU1DVH5LHgZx0fOSCexEmHtJqmZqo8FeJ7AVeAvdF0hByx1NQ1R+Sx4GcdHzkgnsRJh7SapmaqPBXiewFXgL3RdGQCdEgsZVgAAgAEAAIAAAACAAAAAAAMAAAAA
```

Bitcoin Core is able to run `finalizepsbt` on that result, but attempting to broadcast the final transaction with `sendrawtransaction` results in this error:

```
error code: -26
error message:
mandatory-script-verify-flag-failed (Invalid Schnorr signature)
```

If the seed is loaded in a Sparrow wallet, Sparrow is able to sign that same PSBT and broadcast the final transaction. Comparing both final transactions can be seen that Sparrow and embit do not produce the same result byte per byte in the signature section of the transaction, and I believe these should be identical.

These are the two transactions side by side. The first one is Sparrow (broadcastable), the second one Embit's.
```
0200000000010117c2c96206fcd887385ac591bdf8fed95515cfcdd30548190cd642bc748f6a4b0000000000fdffffff01b063cd1d0000000022512097777610ff958523da2ea14070ede03a6680a7074fd2e9537e36e16674241a960141626d225da279392453086d8a2b79e2b9e3ea4afa0e4a4128d9311ea4cbf3b7a3adabf7c7ecdf810c506e40179934437f6e7236af211e71029e3c40d712eadd86834d3e0000
0200000000010117c2c96206fcd887385ac591bdf8fed95515cfcdd30548190cd642bc748f6a4b0000000000fdffffff01b063cd1d0000000022512097777610ff958523da2ea14070ede03a6680a7074fd2e9537e36e16674241a96014184b0a0bed4a59949bc8b2340bc67165c4d76fc4216cc5641b59a2b27c480f08b673cde8d8225e20228fcb558c7ecf8f00d14dc6c523bfe67b61f04c412096d45834d3e0000
```

The following tests have been carried out in a custom signet network. If you want to run the same tests you need to connect Sparrow in signet mode to the `ssl://electrum.bitcoinbarcelona.xyz:60602` electrum server. An [explorer](https://sigpool.bitcoinbarcelona.xyz/) is available.

### LINKED ISSUE #86: Missing PSBT v2 Tests and Support for Required Locktime Fields (0x11, 0x12)
Currently, embit lacks both tests and implementation for the following PSBT v2 input fields as defined in [BIP-370](https://github.com/bitcoin/bips/blob/master/bip-0370.mediawiki):

## PSBT v2 – Missing Fields

| Field                          | Key Type                                  | Value                                                              | Description                                                                                      |
|-------------------------------|--------------------------------------------|--------------------------------------------------------------------|--------------------------------------------------------------------------------------------------|
| Required Time-based Locktime  | `PSBT_IN_REQUIRED_TIME_LOCKTIME (0x11)`   | 32-bit little-endian unsigned integer ≥ `500000000`                | Minimum Unix timestamp the transaction's `nLockTime` must meet or exceed                         |
| Required Height-based Locktime| `PSBT_IN_REQUIRED_HEIGHT_LOCKTIME (0x12)` | 32-bit little-endian unsigned integer < `500000000`                | Minimum block height that must be set as the transaction’s `nLockTime`                          |

## Plan of Action

- [ ] Implement parsing and serialization logic for both locktime fields (`0x11` and `0x12`)
- [ ] **Write tests** covering:
  - Valid and invalid values for each field
  - Edge cases (e.g., exactly 500000000)
  - Combined usage with other PSBT v2 fields
- [ ] Cross-verify behavior against the [BIP-370 test vectors](https://github.com/bitcoin/bips/blob/master/bip-0370.mediawiki#test-vectors)

### LINKED ISSUE #146: psbt: `nSequence=0` is silently rewritten to `0xFFFFFFFF`, producing signatures over the wrong transaction
`InputScope.vin` (`src/embit/psbt.py:200`) uses a truthiness check to apply the BIP-370 default sequence:

```python
return TransactionInput(
    self.txid, self.vout, sequence=(self.sequence or 0xFFFFFFFF)
)
```

`nSequence = 0` is a legal and meaningful value, it signals RBF and *enables* `nLockTime`, but it is falsy in Python, so `0 or 0xFFFFFFFF` yields `0xFFFFFFFF`. Any input whose sequence is `0` is silently rewritten to the final sequence.

Because `PSBT.tx` (line 665) rebuilds the transaction from `inp.vin`, and `sighash_segwit` / `sighash_legacy` / `sighash_taproot` (lines 673–680) delegate to `self.tx`, **signing computes the sighash over the rewritten transaction**. The resulting signature does not commit to the transaction actually contained in the PSBT.

The same pattern on the next line affects the transaction version:

```python
version=self.tx_version or 2,     # line 667: a wire version of 0 becomes 2
```

## Reproduction

Verified on master @ `fff7ffa` (2026-06-01) and on the released 0.8.0.

```python
import struct
from embit.psbt import PSBT
from embit.bip32 import HDKey
from embit.transaction import Transaction, TransactionInput, TransactionOutput
from embit.script import Script, p2wpkh, p2pkh
from embit import ec

def cs(n): return bytes([n]) if n < 253 else b'\xfd' + struct.pack('<H', n)
def kv(k, v): return cs(len(k)) + k + cs(len(v)) + v

root = HDKey.from_seed(b'\x01' * 32)
H = 0x80000000
path = [84 | H, 0 | H, 0 | H, 0, 0]
key = root
for idx in path:
    key = key.derive([idx])
pk = key.key.get_public_key()
spk = p2wpkh(pk)

AMT, SEQ, LOCK = 100000, 0, 500000          # SEQ = 0: RBF-signalling, locktime-enabled
txid = bytes.fromhex('aa' * 32)
out_spk = bytes([0x00, 0x14]) + bytes.fromhex('bb' * 20)

tx  = struct.pack('<i', 2) + cs(1) + txid + struct.pack('<I', 0) + cs(0) + struct.pack('<I', SEQ)
tx += cs(1) + struct.pack('<q', 90000) + cs(len(out_spk)) + out_spk + struct.pack('<I', LOCK)

inmap = kv(bytes([0x01]), struct.pack('<q', AMT) + cs(len(spk.data)) + spk.data) \
      + kv(bytes([0x06]) + pk.sec(),
           root.my_fingerprint + b''.join(struct.pack('<I', i) for i in path))
d = b'psbt\xff' + kv(bytes([0x00]), tx) + b'\x00' + inmap + b'\x00' + b'\x00'

p = PSBT.parse(d)
print("sequence on the wire :", hex(SEQ))
print("p.tx.vin[0].sequence :", hex(p.tx.vin[0].sequence))   # -> 0xffffffff
print("round-trips          :", p.serialize() == d)          # -> False

p.sign_with(root)
sig_der = list(p.inputs[0].partial_sigs.values())[0][:-1]

sc = p2pkh(pk)   # BIP143 scriptCode for P2WPKH
def mk(seq):
    return Transaction(version=2, locktime=LOCK,
        vin=[TransactionInput(txid, 0, Script(b''), sequence=seq)],
        vout=[TransactionOutput(90000, Script(out_spk))])

h_real = mk(0).sighash_segwit(0, sc, AMT)
h_rewritten = mk(0xffffffff).sighash_segwit(0, sc, AMT)
print("valid for the PSBT's own tx :", pk.verify(ec.Signature.parse(sig_der), h_real))
print("valid for the rewritten tx  :", pk.verify(ec.Signature.parse(sig_der), h_rewritten))
```

Output:

```
sequence on the wire : 0x0
p.tx.vin[0].sequence : 0xffffffff
round-trips          : False
valid for the PSBT's own tx : False
valid for the rewritten tx  : True
```

## Suggested fix

Distinguish "absent" from "zero" explicitly:

```python
@property
def vin(self):
    sequence = 0xFFFFFFFF if self.sequence is None else self.sequence
    return TransactionInput(self.txid, self.vout, sequence=sequence)
```

and likewise for the version on line 667:

```python
version=2 if self.tx_version is None else self.tx_version,
```

(`locktime=self.locktime or 0` on line 668 is harmless, since the fallback equals the falsy value, but `is None` there would be more consistent.)

-----------------------

Found it with differential fuzzing of PSBT parsing across implementations (Bitcoin Core, rust-psbt, libwally-core, btcd, NBitcoin, bitcoin-s, embit) in [bitcoinfuzz](https://github.com/bitcoinfuzz/bitcoinfuzz).

### PR DESCRIPTION
**Pull request by odudex**
This PR carves the PSBT work out of #145 (Full Silent Payment Implementation) into a smaller, self-contained branch so it can be reviewed and merged first. It adds PSBTv2 support to `PSBT` and `PSBTView`, then fixes a series of correctness and hardening gaps found while reviewing PSBT handling end to end. Silent payments (BIP-352/374/375/376/392) will be rebased on top of it as a follow-up PR.

Nothing in this PR is silent-payments specific. The only forward-looking piece is `PSBT._sign_taproot_keypath`, a small extraction the SP signer subclasses.

### What is in it

**PSBTv2 (BIP-370)**
- Parsing, validation and serialization of the v2 global, input and output fields in both `PSBT` and `PSBTView`; version-scoped field checks (v2 fields rejected in v0 and vice versa), canonical count fields, required-field checks.
- Unsigned transaction reconstruction, BIP-370 locktime determination from per-input requirements, `PSBT_GLOBAL_TX_MODIFIABLE` updates on signing, Creator/Constructor helpers (`PSBT.create_v2`, `add_input`, `add_output`).
- All BIP-370 test vectors run through both `PSBT` and `PSBTView`; valid vectors must round-trip byte for byte through both.

**Correctness fixes**
- Taproot sighash with `ANYONECANPAY` hashed the whole serialized input instead of the outpoint, and `SIGHASH_SINGLE` hashed the output instead of its SHA256, in `Transaction` and `PSBTView` (#65). Verified against every hash type in the BIP-341 wallet test vectors.
- `nSequence = 0` was rewritten to `0xffffffff` and `tx_version = 0` to `2` before signing (#146), in `InputScope`, `PSET` and `PSET.blinded_tx`.
- `PSET` signed PSETv2 with the fallback locktime instead of the BIP-370 derived one.
- `PSBTView` matched v2 prevout/output fields by key prefix, so a keyed or duplicated field could feed a wrong value into `hashPrevouts`/`hashOutputs` and a signature could be emitted before the malformed input was parsed.
- `PSBT.parse` rejected PSBTv0 whose unsigned transaction has no inputs (#117).
- `PSBT` and `PSBTView` now produce identical bytes: same v2 global key order, both store `PSBT_IN_TAP_KEY_SIG` on taproot key-path spends, same locktime API (`.locktime` is the raw field, `determine_locktime()` the derived one).

**Hardening**
- Every malformed PSBT raises `PSBTError` from both parsers: truncated input, non-canonical or truncated compact sizes (`compact.read_from` now matches Bitcoin Core), corrupted nested transactions, scripts and keys, absurd lengths, trailing bytes, out-of-range values on serialization. Tests truncate and bit-flip sample PSBTs at every byte.
- `PSBT_IN_NON_WITNESS_UTXO` is held to its declared length so a wrong length cannot shift the map framing.
- `PSBTView.view()` walks all maps once and rejects trailing data and count/map mismatches, like `PSBT.parse`.
- Inputs without a utxo, PSET outputs without an asset, and duplicate keys anywhere in a map are rejected with `PSBTError`.

**PSBTView performance**
- Prevouts, sequences and outputs are hashed in one pass; taproot utxo data is cached per view. Signing 40 inputs went from 33,580 scope skips to 2,380 (p2pkh v2) and from 1,680 input-scope parses to 80 (taproot).

### Behaviour changes to be aware of
- `PSBTView.locktime` now means the same as `PSBT.locktime`: the raw fallback field for v2 (`None` when absent). Callers that displayed the derived transaction locktime from a view should use `determine_locktime()`.
- `PSBTView.view()` rejects a stream with data after the PSBT, as `PSBT.parse` always did.
- `compact.read_from` rejects non-canonical encodings, as Bitcoin Core does.
- `PSBT.parse` raises `PSBTError` (still an `EmbitError`) instead of a bare `EmbitError` on trailing bytes.

### Issues covered
- Closes #86 (full BIP-370 support).
- Closes #146 (`nSequence=0` rewritten before signing).
- Closes #65 (invalid taproot signatures with non-default sighash).
- #80 (explicit `SIGHASH_ALL` byte on Schnorr signatures): embit appends the byte only when the caller asks for `SIGHASH.ALL` explicitly, which is what BIP-341 specifies. With the default `SIGHASH.DEFAULT` the signature is 64 bytes. Left as is, the fix belongs in the caller; noted here so it can be closed if that is agreed.

### Authorship
- The PSBTv2 implementation itself (`feat(BIP-370): PSBTv2 support`, the global-field validation and parsing fixes, the `TX_MODIFIABLE` dedupe and the `_sign_taproot_keypath` extraction) is authored by @notTanveer and cherry-picked from #145 with authorship preserved.
- The BIP-341 sighash fix credits @1ma (Marcel Hernandez) as co-author; it is the fix from #66.
- The zero-input unsigned transaction fix credits @sinancrypto (Sinan) as co-author; it is the approach from #117.
- The remaining hardening, parity, performance and test commits are new in this PR.

### PRs covered or replaced
- #87 (PSBTv2 support, @notTanveer): superseded by this PR, which contains its rebased content plus the fixes above.
- #66 (sighash_taproot fixes, @1ma): the same two fixes, with BIP-341 vector tests; @1ma is credited as co-author of that commit. Can be closed.
- #117 (zero-input unsigned transactions, @sinancrypto): the same approach, `Transaction.read_from(segwit=)`, with tests; @sinancrypto is credited as co-author of that commit. Can be closed.
- #149 (PSBTv0 missing `PSBT_GLOBAL_UNSIGNED_TX`): already rejected on master since #150; this PR keeps that check in the shared validation. Can be closed.
- #145 (Full Silent Payment Implementation, @notTanveer): this PR is its PSBT half. #145 will be rebased on top once this is merged.
- #150 (align version handling with BIP 174/370): merged, this PR builds on it.

### CODE REVIEW

**Comment by odudex** [comment]
> Locktime determination can incorrectly accept mixed (height+time) locktime requirements (including the case where an input specifies both), which should be rejected to avoid deriving an invalid signing locktime.

Rejecting the "both" case would fail the first two BIP vectors and diverge from Bitcoin Core, whose ComputeTimeLock treats an input carrying both fields as satisfiable by either type.

File: src/embit/psbt.py [hunk 1]
```diff
@@ -437,7 +664,7 @@ def read_value(self, stream, k):
             pub = ec.PublicKey.from_xonly(k[1:])
             if pub not in self.taproot_bip32_derivations:
                 b = BytesIO(v)
-                num_leaf_hashes = compact.read_from(b)
+                num_leaf_hashes = read_compact(b)
                 leaf_hashes = [b.read(32) for i in range(num_leaf_hashes)]
                 if not all([len(leaf) == 32 for leaf in leaf_hashes]):
                     raise PSBTError("Invalid length of taproot leaf hashes")
```
Comment (notTanveer) [inline src/embit/psbt.py:669]: check before allocating memory to prevent OOM attacks

```suggestion
                num_leaf_hashes = read_compact(b)
                if num_leaf_hashes * 32 > len(v) - b.tell():
                    raise PSBTError("Invalid length of taproot leaf hashes")
                leaf_hashes = [b.read(32) for i in range(num_leaf_hashes)]
```

File: src/embit/psbt.py [hunk 2]
```diff
@@ -633,7 +936,7 @@ def read_value(self, stream, k):
...
```
Comment (notTanveer) [inline src/embit/psbt.py:941]:
```suggestion
                num_leaf_hashes = read_compact(b)
                if num_leaf_hashes * 32 > len(v) - b.tell():
                    raise PSBTError("Invalid length of taproot leaf hashes")
                leaf_hashes = [b.read(32) for i in range(num_leaf_hashes)]
```

Comment (odudex) [reply src/embit/psbt.py:941]: Thank you @notTanveer ! Both suggestions and a test added as a new commit, co-authoring you.
